File: 01-content-factory/integrations/elevenlabs.ts
const ELEVENLABS_BASE_URL = "https://api.elevenlabs.io/v1";

export interface VoiceoverResult {
  success: boolean;
  audioUrl?: string;
  audioData?: Buffer;
  duration?: number;
  error?: string;
}

export interface Voice {
  voice_id: string;
  name: string;
  category: string;
  labels?: Record<string, string>;
}

const DEFAULT_VOICES: Record<string, string> = {
  professional_male: "pNInz6obpgDQGcFmaJgB", // Adam
  professional_female: "21m00Tcm4TlvDq8ikWAM", // Rachel
  warm_male: "VR6AewLTigWG4xSOukaG", // Arnold
  warm_female: "EXAVITQu4vr4xnSDxMaL", // Bella
  energetic: "ErXwobaYiN019PkySvjV", // Antoni
  calm: "MF3mGyEYCl7XYWbV9V6O", // Elli
};

export async function generateVoiceover(
  text: string,
  options: {
    voiceId?: string;
    voiceStyle?: keyof typeof DEFAULT_VOICES;
    stability?: number;
    similarityBoost?: number;
    speed?: number;
  } = {}
): Promise<VoiceoverResult> {
  const apiKey = process.env.ELEVENLABS_API_KEY;
  
  if (!apiKey) {
    return {
      success: false,
      error: "ELEVENLABS_API_KEY not configured. Add it to your secrets.",
    };
  }

  const {
    voiceId = DEFAULT_VOICES.professional_male,
    voiceStyle,
    stability = 0.5,
    similarityBoost = 0.75,
    speed = 1.0,
  } = options;

  const selectedVoiceId = voiceStyle ? DEFAULT_VOICES[voiceStyle] : voiceId;

  try {
    const response = await fetch(
      `${ELEVENLABS_BASE_URL}/text-to-speech/${selectedVoiceId}`,
      {
        method: 'POST',
        headers: {
          'xi-api-key': apiKey,
          'Content-Type': 'application/json',
          'Accept': 'audio/mpeg',
        },
        body: JSON.stringify({
          text,
          model_id: "eleven_multilingual_v2",
          voice_settings: {
            stability,
            similarity_boost: similarityBoost,
            speed,
          },
        }),
      }
    );

    if (!response.ok) {
      const errorText = await response.text();
      console.error("[ElevenLabs] API error:", response.status, errorText);
      return {
        success: false,
        error: `ElevenLabs API error: ${response.status} - ${errorText}`,
      };
    }

    const audioBuffer = Buffer.from(await response.arrayBuffer());
    
    const estimatedDuration = Math.ceil(text.split(/\s+/).length / 2.5);

    console.log(`[ElevenLabs] Generated ${audioBuffer.length} bytes of audio (~${estimatedDuration}s)`);

    return {
      success: true,
      audioData: audioBuffer,
      duration: estimatedDuration,
    };
  } catch (error: any) {
    console.error("[ElevenLabs] Voiceover error:", error);
    return {
      success: false,
      error: error.message || "Failed to generate voiceover",
    };
  }
}

export async function generateVoiceoverWithUrl(
  text: string,
  options: {
    voiceId?: string;
    voiceStyle?: keyof typeof DEFAULT_VOICES;
  } = {}
): Promise<VoiceoverResult> {
  const apiKey = process.env.ELEVENLABS_API_KEY;
  
  if (!apiKey) {
    return {
      success: false,
      error: "ELEVENLABS_API_KEY not configured",
    };
  }

  const result = await generateVoiceover(text, options);
  
  if (!result.success || !result.audioData) {
    return result;
  }

  const base64Audio = result.audioData.toString('base64');
  const audioUrl = `data:audio/mpeg;base64,${base64Audio}`;

  return {
    success: true,
    audioUrl,
    duration: result.duration,
  };
}

export async function getAvailableVoices(): Promise<Voice[]> {
  const apiKey = process.env.ELEVENLABS_API_KEY;
  
  if (!apiKey) {
    console.warn("[ElevenLabs] API key not configured, returning default voices");
    return Object.entries(DEFAULT_VOICES).map(([name, id]) => ({
      voice_id: id,
      name: name.replace(/_/g, ' '),
      category: 'default',
    }));
  }

  try {
    const response = await fetch(`${ELEVENLABS_BASE_URL}/voices`, {
      headers: {
        'xi-api-key': apiKey,
      },
    });

    if (!response.ok) {
      console.error("[ElevenLabs] Failed to fetch voices:", response.status);
      return [];
    }

    const data = await response.json();
    return data.voices || [];
  } catch (error) {
    console.error("[ElevenLabs] Error fetching voices:", error);
    return [];
  }
}

export async function estimateAudioDuration(text: string): Promise<number> {
  const words = text.split(/\s+/).length;
  return Math.ceil(words / 2.5);
}

export { DEFAULT_VOICES };

File: 01-content-factory/python/providers/elevenlabs.py
"""
ElevenLabs Voice AI Provider for voiceovers
"""
import os
import httpx
from typing import Optional


class ElevenLabsProvider:
    """ElevenLabs provider for AI voiceovers"""
    
    BASE_URL = "https://api.elevenlabs.io/v1"
    
    def __init__(self):
        self.api_key = os.environ.get("ELEVENLABS_API_KEY")
        self.available = bool(self.api_key)
        self.default_voice_id = os.environ.get("ELEVENLABS_VOICE_ID", "21m00Tcm4TlvDq8ikWAM")
        
        if not self.available:
            print("ELEVENLABS_API_KEY not set - ElevenLabs provider disabled")
    
    async def generate_voiceover(
        self,
        text: str,
        voice_id: Optional[str] = None,
        model_id: str = "eleven_turbo_v2_5",
    ) -> Optional[bytes]:
        """Generate voiceover audio from text"""
        if not self.available:
            print("ElevenLabs not available - returning mock")
            return None
        
        voice = voice_id or self.default_voice_id
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    f"{self.BASE_URL}/text-to-speech/{voice}",
                    headers={
                        "xi-api-key": self.api_key or "",
                        "Content-Type": "application/json",
                    },
                    json={
                        "text": text,
                        "model_id": model_id,
                        "voice_settings": {
                            "stability": 0.5,
                            "similarity_boost": 0.75,
                        },
                    },
                    timeout=60.0,
                )
                
                if response.status_code == 200:
                    return response.content
                else:
                    print(f"ElevenLabs error: {response.status_code} - {response.text}")
                    return None
                    
            except Exception as e:
                print(f"ElevenLabs API error: {e}")
                return None
    
    async def get_voices(self) -> list:
        """Get list of available voices"""
        if not self.available:
            return []
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(
                    f"{self.BASE_URL}/voices",
                    headers={"xi-api-key": self.api_key or ""},
                )
                
                if response.status_code == 200:
                    return response.json().get("voices", [])
                return []
                
            except Exception as e:
                print(f"ElevenLabs API error: {e}")
                return []
